TRUNK-6208:Upgrade 2.7 Platform to Liquibase 4.26.0

Seremba opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'TRUNK-5111: Replace use of deprecated isVoided' -->
<!--- 'TRUNK-JiraIssueNumber: JiraIssueTitle' -->
## Description of what I changed
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->

## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first! -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it! -->
<!--- Just add the issue number at the end: -->
see https://openmrs.atlassian.net/browse/TRUNK-6208

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

Seremba commented:
Hi @dkayiwa, when I run `mvn clean install` after changing the version of  `liquibase-core` and `liquibase-plugin` in pom.xml to 4.26.0, I get this error https://pastebin.com/aGPRUqu9. According to the liquibase [docs](https://docs.liquibase.com/concepts/changelogs/how-liquibase-finds-files.html), from version 4.13.0, we need to include `search-path` property for liquibase to find its files. I thought of liquibase.properties but I don't seem to find something like that in OpenMRS core. Kindly point me to that place where I can configure the search-path. Thanks a lot.

dkayiwa commented:
> search-path



> Hi @dkayiwa, when I run `mvn clean install` after changing the version of `liquibase-core` and `liquibase-plugin` in pom.xml to 4.26.0, I get this error https://pastebin.com/aGPRUqu9. According to the liquibase [docs](https://docs.liquibase.com/concepts/changelogs/how-liquibase-finds-files.html), from version 4.13.0, we need to include `search-path` property for liquibase to find its files. I thought of liquibase.properties but I don't seem to find something like that in OpenMRS core. Kindly point me to that place where I can configure the search-path. Thanks a lot.

Where did the docs say that the `search-path` should be configured from?

tendomart commented on api/src/main/java/org/openmrs/util/DatabaseUpdater.java:
```diff
@@ -233,8 +231,8 @@ public static List<String> executeChangelog(String changeLogFile, Contexts conte
 			//
-			ChangeLogHistoryServiceFactory.getInstance().getChangeLogService(database).init();
-			
+			// ChangeLogHistoryServiceFactory.getInstance().getChangeLogService(database).init();
```
Remove commented out code

Seremba commented:
> > search-path
> 
> > Hi @dkayiwa, when I run `mvn clean install` after changing the version of `liquibase-core` and `liquibase-plugin` in pom.xml to 4.26.0, I get this error https://pastebin.com/aGPRUqu9. According to the liquibase [docs](https://docs.liquibase.com/concepts/changelogs/how-liquibase-finds-files.html), from version 4.13.0, we need to include `search-path` property for liquibase to find its files. I thought of liquibase.properties but I don't seem to find something like that in OpenMRS core. Kindly point me to that place where I can configure the search-path. Thanks a lot.
> 
> Where did the docs say that the `search-path` should be configured from?
From the project’s classpath if we are using Maven. cc @dkayiwa

Seremba commented:
> > search-path
> 
> > Hi @dkayiwa, when I run `mvn clean install` after changing the version of `liquibase-core` and `liquibase-plugin` in pom.xml to 4.26.0, I get this error https://pastebin.com/aGPRUqu9. According to the liquibase [docs](https://docs.liquibase.com/concepts/changelogs/how-liquibase-finds-files.html), from version 4.13.0, we need to include `search-path` property for liquibase to find its files. I thought of liquibase.properties but I don't seem to find something like that in OpenMRS core. Kindly point me to that place where I can configure the search-path. Thanks a lot.
> 
> Where did the docs say that the `search-path` should be configured from?

From the project’s classpath if we are using Maven. cc @dkayiwa

Seremba commented:
Hi @dkayiwa when you happen to find time, kindly review this PR!

Seremba commented:
Hi @dkayiwa I have added more changes to this PR, kindly review. Thanks a lot!

dkayiwa commented on api/src/test/resources/org/openmrs/util/testLiquibase.xml:
```diff
@@ -23,9 +23,10 @@
+
 	<changeSet id="testSqlFile" author="OpenMRS" runAlways="true">
-		<sqlFile endDelimiter=";" path="sql/testSqlFile.sql" relativeToChangelogFile="true" stripComments="true" />
+		<sqlFile endDelimiter=";" path="org/openmrs/util/sql/testSqlFile.sql" stripComments="true"
```
Do you mind explaining why the path was changed?

dkayiwa commented on api/src/test/java/org/openmrs/liquibase/OpenmrsClassLoaderResourceAccessorTest.java:
```diff
@@ -51,10 +44,8 @@ public void shouldGetNoResourceAsStream() throws Exception {
-			 InputStreamList inputStreamSet = classLoaderFileOpener.openStreams(null, "")){
-			assertThat(inputStreamSet.size(), is(0));
-		}
+		OpenmrsClassLoaderResourceAccessor classLoaderFileOpener = new OpenmrsClassLoaderResourceAccessor(classLoader);
```
Why did you remove the `try`?

Seremba commented:
Hi @dkayiwa, do you have any idea why the test fail? I tested the upgrade this way: 
1. I `cd` into `openmrs-core`
2. deleted the OpenMrs folder.
3. then `cd` into `openmrs-core/webapp`
4. went to http://localhost:8080/openmrs/initialsetup
5. chose simple, entered my db password and it successfully created and updated the database. Is there anything I am doing wrong or I am meant do? Thanks a lot

dkayiwa commented:
Feel free to share this on talk.

Seremba commented:
> Feel free to share this on talk.

thanks a lot!

Seremba commented:
Hi @dkayiwa, added more changes to this PR. Kindly review! Liquibase 4.27.0 has been released less than a week ago. So I decided to update to it instead of 4.26.0. I hope that is okay!

dkayiwa commented:
@Seremba did you successfully run `mvn clean install`?

Seremba commented:
@dkayiwa `mvn clean install` successfully runs. However, I have noticed that after v4.21.1, the app fails to start. It throws this error
https://pastebin.com/94gBF88S. When I upgrade to versions greater than 4.21.1, it(the version) throws deprecation warnings mainly on two API methods that we use:
1) 'listUnrunChangeSets(liquibase.Contexts, liquibase.LabelExpression)' is deprecated. Unfortunately, Liquibase docs do not provide an alternative to this method.
2) 'ChangeLogHistoryServiceFactory.getInstance()' has an alternative according to the docs. 
All the changelog files are quite similar, but this `liquibase-schema-only-1.9.x.xml` has a `<modifySql>` tag. I wonder if this could be the reason for the validation failures.
Searching around the web suggests clearing checksums, but I wouldn't like to clear a checksum that has been around since 2011. 
My suggestion is we upgrade up to 4.21.1 at least for now.

Seremba commented:
@dkayiwa the above commit is a test commit for what happens if we update to 4.21.1

Seremba commented:
@dkayiwa with all honesty, I didn't expect this to fail. As everything works well on locally!

dkayiwa commented:
Run this command: `mvn test -Pskip-default-test -Pintegration-test`

Seremba commented:
Thank you, yeah that fails. It also tells why and what to do!

dkayiwa commented:
When you look at the build failure log, after clicking the `Details` link, you get to see the actual command that is being run, at the top of the log. That is where i got the command that i shared above. Does this make sense? 😊

Seremba commented:
> When you look at the build failure log, after clicking the `Details` link, you get to see the actual command that is being run, at the top of the log. That is where i got the command that i shared above. Does this make sense? 😊

Yeah sure I do!!! Thanks for the tip. Let me get back to this when the power is back. Battery so low

dkayiwa commented:
@Seremba i have unassigned you from this ticket because its complexity does not fit your current level and we need it done soonish for the next platform release. Feel free to pick any ticket with the intro label and of low complexity.

Seremba commented:
Oh okay! I will be happy learn on how it was solved as well. Thanks @dkayiwa
